include/pybind11/detail/type_caster_base.h

```diff
@@ -827,6 +827,33 @@ using movable_cast_op_type
 template <typename T, typename SFINAE = void>
 struct is_copy_constructible : std::is_copy_constructible<T> {};
 
+// True if Container has a dependent type mapped_type that is equivalent
+// to Container itself
+// Actual implementation in the SFINAE specializations below
+template <typename Container, typename MappedType = Container>
+struct sfinae_is_container_with_self_referential_mapped_type : std::false_type {};
+
+// Tie-breaking between the mapped_type and the value_type specializations is trivial:
+// The specializations are only valid if both conditions are fulfilled:
+// 1) The mapped_type (respectively value_type) exists
+// 2) And it is equivalent to Container
+// So, in each case, only one specialization will activate.
+template <typename Container>
+struct sfinae_is_container_with_self_referential_mapped_type<Container,
+                                                             typename Container::mapped_type>
+    : std::true_type {};
+
+template <typename Container>
+struct sfinae_is_container_with_self_referential_mapped_type<Container,
+                                                             typename Container::value_type>
+    : std::true_type {};
+
+// Use a helper struct in order to give this a nicer public API without helper template parameter.
+// This makes this struct nicer to specialize by users.
+template <typename Container>
+struct is_container_with_self_referential_mapped_type
```

Comment:
Not sure, but tossing this around in my mind, I feel

`is_recursive_container`

would be great. It takes a small mental leap back to "self-referential", vs a small leap forward from "self-referential" to "recursive". "recursive" is shorter, a little more intuitive at least in my mind, and we don't have the trouble with `mapped_type` vs `value_type`, which is is more-or-less a case-specific detail.

Reply:
:+1: